src/classes/RD2_OpportunityService.cls

```diff
@@ -0,0 +1,210 @@
+
+    /***
+    * @description inserts one or more Opportunity records for a specified set of Recurring Donations
+    * @param rdsWithCloseDates the Map of Recurring Donations with close dates
+    */
+    public void createInstallmentOpportunities(Map<npe03__Recurring_Donation__c, List<Date>> rdsWithCloseDates) {
+
+        List<Opportunity> oppsToInsert;
+        Set<Id> createdOppIds = new Set<Id>();
+        Integer oppItem = 0;
+        //Create a list of ContactIds and contactmap used in these RecurringDonations
+        List<Id> contactIds = new List<Id>();
+        Map<Id,Contact> contactMap;
+        Map<Id,String> rdPaymentMethod = new Map<Id,String>();
+        List<Database.SaveResult> badSRs = new List<Database.SaveResult>();
+        List<Opportunity> badOpps = new List<Opportunity>();
+
+        for (npe03__Recurring_Donation__c r : rdsWithCloseDates.keyset()) {
+            if (r.npe03__Contact__c != null) {
+                contactIds.add(r.npe03__Contact__c);
+            }
+            rdPaymentMethod.put(r.Id,r.PaymentMethod__c);
+
+        }
+        contactMap = getContacts(contactIds);
+
+        oppsToInsert = getInstallmentOpportunities(rdsWithCloseDates, contactMap);
+
+        if (oppsToInsert.size() > 0) {
+            for (Database.SaveResult s : insertOpportunities(oppsToInsert)) {
+                if (s.isSuccess() == true) {
+                    //get all the new OppIds into a list
+                    createdOppIds.add(s.getId());
+
+                } else {//handle errors on insert
+                    badSRs.add(s);
+                    badOpps.add(oppsToInsert[oppItem]);
+                }
+                oppItem++;
+            }
+            //Confirm if the error handling is needed due to bulk operations
+            if (!badSRs.isEmpty()) {
+               ERR_Handler.Errors errors = ERR_Handler.getErrors(badSRs, badOpps);
+               ERR_Handler.processErrors(errors.errorRecords, ERR_Handler_API.Context.RD.name());
+            }
+        }
+        updatePaymentRecordOnCreatedOpp(createdOppIds, rdPaymentMethod);
+    }
+
+    /***
+    * @description Gets one or more Opportunity records for a specified set of Recurring Donations with close dates
+    * @param rdsWithCloseDates the Map of Recurring Donations with close dates
+    * @param contactMap the Map of contact with account ids.
+    * @return the list of Opportunities to be created.
+    */
+    private List<Opportunity> getInstallmentOpportunities(Map<npe03__Recurring_Donation__c, List<Date>> rdsWithCloseDates, Map<Id,Contact> contactMap) {
+
+        List<Opportunity> oppsToInsert = new List<Opportunity>();
+        npe03__Recurring_Donations_Settings__c rds = UTIL_CustomSettingsFacade.getRecurringDonationsSettings();
+
+        for (npe03__Recurring_Donation__c r : rdsWithCloseDates.keySet()) {
+
+            for(Integer i = 0;i< rdsWithCloseDates.get(r).size();i++) {
+                Opportunity opp = new Opportunity();
```

Comment:
I think this loop might be easier to read if you abstracted out th elogic to create a new Opp into a new method so that the loop is basically one line:
`oppsToInsert.add(buildOppForRd(rd, rdsWithCloseDates.get(r)[i]);`

```
private Opportunity buildOppForRd(npe03__Recurring_Donation__c rd, Date closeDate) {
   Opportunity opp = new Opportuntity(
        AccountId = rd.npe03__Organization__c,
        Primary_Contact__c rd.npe03__Contact__c,
        CloseDate = closeDate,
        StageName = rdSettings.InstallmentOppStageName__c,
        Amount = rd.npe03__Amount__c,
        npe03__Recurring_Donation__c = r.dId,
        Recurring_Donation_Installment_Number__c = (rd.npe03__Total_Paid_Installments__c != null ?
    npe03__Total_Paid_Installments__c + 1 : 1)
    );
    if (rd.npe03__Recurring_Donation_Campaign__c != null && rdSettings.npe03__Add_Campaign_to_All_Opportunites__c ) {
        opp.CampaignId = rd.npe03__Recurring_Donation_Campaign__c;
    }
    return opp;
}
```